### DISCUSSION TITLE
Does Uninstall EfiSimpleNetwork protocol happen?

### DISCUSSION THREAD

**Discussion by amdhugo** [ipxe/ipxe#401] [Q&A]
In one of previous discussion, you mentioned 
iPXE does uninstall the EfiSimpleNetwork protocol. The MNP driver should react to this by stopping polling the SNP device. If the MNP driver does not do this then there is a bug in the MNP driver.
As of commit 5b41b9a (from October 2020) iPXE will work around this kind of bug in the MNP driver (or other consumers of protocols published by iPXE): if the call to UninstallMultipleProtocolInterfaces reports an error due to a buggy BIOS driver then iPXE will deliberately leave the protocol present but disabled, leaking the memory since it is not safe to free the protocol.

However, from the team's investigation, we might not figure out that iPXE unintalls EfiSimpleNetwork? Would you please confirm what is wrong on that?
 
![image](https://user-images.githubusercontent.com/64545075/123745411-d9031b00-d8e2-11eb-8b91-1228f28492f5.png)

**Comment by mcb30** [comment]
What message were you expecting to see?  There is no debug message printed for a successful uninstallation, since it is an unremarkable event.

Also, the `Richard ...` debug message text does not appear anywhere in the codebase, which indicates that you are using a modified source code.  Please either use the unmodified upstream source code, or explain precisely what changes you have made.  You cannot expect us to debug code that we cannot see.

**Comment by mcb30** [comment]
> In one of previous discussion, you mentioned

(This previous discussion was #281)

**Comment by RichardDai123** [comment]
Hi Michael，
 I am richard, I am checking this issue on amd platform with AMI BIOS. the attach file :ipxe_debug.zip is the ipxe src ,and BIOS changes. 
 and a ipxe boot log. you can check it.
[Ipxe_debug.zip](https://github.com/ipxe/ipxe/files/6768917/Ipxe_debug.zip)

To duplicate this issue. I do follow it.
    1.	Downlod the ipxe src with command: git clone git://git.ipxe.org/ipxe.git under ubuntu.
    2.	Edit a script to break the ipxe infinite loop .
    3.	Build the ipxe,efi and open the serial debug mode. Command: make bin-x86_64-efi/ipxe.efi EMBED=demo.ipxe DEBUG=efi_snp,snp,netdevice,efi_driver,snpnet
    4.	Try to do ipxe boot.
    It has the log：ipxe_boot.log.

To check this ipxe_boot,log. You can search keyword: ”Richard” or “ipxe.efi”.

As the log found.
1.The EfiSimpleNetwork has been installed, but not uninstall until shutdown exit

> Line: 1053
> [33mNETDEV net0 registered (phys 0000:03:00.0 hwaddr 00:00:1a:1d:b1:cd)
> 
> [0mRichard efi_snp_probe 2
> [34mRichard efi_snp_probe 1
> [0mRichard install Snp network 
> **InstallProtocolInterface: EfiSimpleNetwork B02BDC50**
> InstallProtocolInterface: EfiDevicePath B02BDA08
> InstallProtocolInterface: E18541CD-F755-4F73-928D-643C8A79B229 B02BE080
> InstallProtocolInterface: EfiNetworkInterfaceIdentifierProtocolGuid_31 B02BE080

2.The ipxe.efi has start ,but not stop well when shutdown exit.

> line :985
> ipxe.efi.Start(B023E1FB)[PciRoot(0x0)/Pci(0x2,0x1)/Pci(0x0,0x2)/Pci(0x8,0x0)/Pci(0x0,0x0)]= **B2CC5A18**

> line 4305
> Richard driver disconnect PciRoot(0x0)/Pci(0x2,0x1)/Pci(0x0,0x2)/Pci(0x8,0x0)/Pci(0x0,0x0) handle:**0xb2cc5a18** 
> [0mRichard DriverImageHandle B0D19B98=
> Richard DriverImageHandle B0D19B98 controllehandle=B1444B18,OpenData->Attributes=8
> Richard DriverImageHandle B0D19B98 DriverImageHandleValid=0   **<=The driver image is not valid. no open protocol with Attributes ”EFI_OPEN_PROTOCOL_BY_DRIVER” under this handle.**

else:
  I do a test ,if mask the line: efipci_close ( device ) in routine "efipci_info"  of ipxe\src\interface\efi\efi_pci.c.
the ipxe.efi can be stopped, and EfiSimpleNetwork uninstalled.

Thanks
Richard Dai

**Reply by NiKiZe** [reply]
Changes to the source is best provided via tools such as git diff. Uploading a full version of sources generally creates to much work, which makes such reports mostly ignored.

**Reply by mcb30** [reply]
> Changes to the source is best provided via tools such as git diff. Uploading a full version of sources generally creates to much work, which makes such reports mostly ignored.

Agreed.  @RichardDai123 please provide a GitHub fork that includes whatever modifications you have made for debugging.  

Please make sure to strip out any irrelevant modifications from your code.  When you are asking for free support from an open-source project (particularly if you are a customer such as AMD who has a record of failing to pay previous bills for iPXE commercial support), you should try to minimise the amount of work that you are asking other people to do for free.

**Reply by amdhugo** [reply]
Is this issue fixed? Anyone who can share the information.

**Reply by NiKiZe** [reply]
> Is this issue fixed? Anyone who can share the information.

Did you read what we wrote? Did you upload the modified sources so that it can actually be used to see what is going on?

**Reply by amdhugo** [reply]
> > Is this issue fixed? Anyone who can share the information.
> 
> Did you read what we wrote? Did you upload the modified sources so that it can actually be used to see what is going on?

I see the comments of this commit and I guess it is fixed, hence, I asked.
commit e09e1142a3bd8bdb702efc92994c419a53e9933b (HEAD -> master, origin/master, origin/coverity_scan, origin/HEAD)
Author: Michael Brown <mcb30@ipxe.org>
Date:   Fri Jul 23 11:32:04 2021 +0100

    [efi] Record cached ProxyDHCPOFFER and PXEBSACK, if present

    Commit cd3de55 ("[efi] Record cached DHCPACK from loaded image's
    device handle, if present") added the ability for a chainloaded UEFI
    iPXE to reuse an IPv4 address and DHCP options previously obtained by
    a built-in PXE stack, without needing to perform a second DHCP
    request.

    Extend this to also record the cached ProxyDHCPOFFER and PXEBSACK
    obtained from the EFI_PXE_BASE_CODE_PROTOCOL instance installed on the
    loaded image's device handle, if present.

    This allows a chainloaded UEFI iPXE to reuse a boot filename or other
    options that were provided via a ProxyDHCP or PXE boot server
    mechanism, rather than by standard DHCP.

**Reply by NiKiZe** [reply]
> I see the comments of this commit and I guess it is fixed, hence, I asked.
> 
> ```
> [efi] Record cached ProxyDHCPOFFER and PXEBSACK, if present
> ```

Not sure why you would think this is related?

**Reply by amdhugo** [reply]
> > I see the comments of this commit and I guess it is fixed, hence, I asked.
> > ```
> > [efi] Record cached ProxyDHCPOFFER and PXEBSACK, if present
> > ```
> 
> Not sure why you would think this is related?

You probably can reply if they are related or not. I don't know if they are, that is why I asked.

**Reply by mcb30** [reply]
That commit is entirely unrelated.

As requested above: please provide a GitHub fork that includes whatever modifications you have made for debugging, and please make sure to strip out any irrelevant modifications from your code.

**Reply by amdhugo** [reply]
> That commit is entirely unrelated.
> 
> As requested above: please provide a GitHub fork that includes whatever modifications you have made for debugging, and please make sure to strip out any irrelevant modifications from your code.

Thanks Michael, There is must misunderstanding. I was told the branch cached_proxydhcp fixes this issue. The latest comments are same both on master and the branch. So It doesn't look like it is.

**Reply by NiKiZe** [reply]
> Thanks Michael, There is must misunderstanding. I was told the branch cached_proxydhcp fixes this issue. The latest comments are same both on master and the branch. So It doesn't look like it is.

Thanks for clarifying, Just to be clear, proxydhcp has nothing to do with SNP.
